test/exercises/diffusion/heat_2d_dist_buffers.chpl

```diff
@@ -0,0 +1,149 @@
+             alpha = 0.25;  // diffusion constant
+
+// define distributed domains and block-distributed array
+const indices = {0..<nx, 0..<ny},
+      indicesInner = indices.expand(-1),
+      INDICES = Block.createDomain(indices);
+var u: [INDICES] real;
+
+// apply initial conditions
+u = 1.0;
+u[nx/4..nx/2, ny/4..ny/2] = 2.0;
+
+// a type for creating a "skyline" array of halo buffers
+record haloArray {
+  var d: domain(1);
+  var v: [d] real;
+
+  proc init() do this.d = {0..0};
+  proc init(r: range(int)) do
+    this.d = {r};
+}
+
+// set up array of halo buffers over same distribution as 'u.targetLocales'
+var LOCALE_DOM = Block.createDomain(u.targetLocales().domain);
+var haloArrays: [LOCALE_DOM] [0..<4] haloArray;
+
+// buffer edge indices: North, East, South, West
+param N = 0, S = 1, E = 2, W = 3;
+
+// number of tasks per dimension based on Block distribution's decomposition
+const tidXMax = u.targetLocales().dim(0).high,
+      tidYMax = u.targetLocales().dim(1).high;
+
+// barrier for one task per locale
+var b = new barrier(numLocales);
+
+proc main() {
+  if RunCommDiag then startCommDiagnostics();
+
+  // spawn one task for each locale
+  t.start();
+  coforall (loc, (tidX, tidY)) in zip(u.targetLocales(), LOCALE_DOM) {
+    // run initialization and computation on the task for this locale
+    on loc {
+      // initialize halo arrays
+      haloArrays[tidX, tidY][N] = new haloArray(u.localSubdomain().dim(1).expand(1));
+      haloArrays[tidX, tidY][S] = new haloArray(u.localSubdomain().dim(1).expand(1));
+      haloArrays[tidX, tidY][E] = new haloArray(u.localSubdomain().dim(0).expand(1));
+      haloArrays[tidX, tidY][W] = new haloArray(u.localSubdomain().dim(0).expand(1));
+
+      // synchronize across tasks
+      b.barrier();
+
+      // run the portion of the FD computation owned by this task
+      work(tidX, tidY);
+    }
+  }
+
+  if RunCommDiag {
+    stopCommDiagnostics();
+    printCommDiagnosticsTable();
+  }
+
+  // print final results
+  const mean = (+ reduce u) / u.size,
+        stdDev = sqrt((+ reduce (u - mean)**2) / u.size);
+  t.stop();
+  writeln(abs(0.222751 - stdDev) < 1e-6);
+  if writeTime then writeln("time: ", t.elapsed());
+}
+
+proc work(tidX: int, tidY: int) {
+  // define domains to describe the indices owned by this task
+  const localIndices = u.localSubdomain(here),
+        localIndicesBuffered = localIndices.expand(1),
+        localIndicesInner = localIndices[indicesInner];
+
+	// declare two local arrays for computation on this locale
+  var uLocal1, uLocal2: [localIndicesBuffered] real = 1;
+
+  // populate local arrays with initial conditions from global array
+  uLocal1[localIndices] = u[localIndices];
+  uLocal2 = uLocal1;
+
+  // define constants for indexing into edges of local array
+  const WW = localIndicesBuffered.dim(1).low,
+        EE = localIndicesBuffered.dim(1).high,
+        SS = localIndicesBuffered.dim(0).high,
+        NN = localIndicesBuffered.dim(0).low;
```

Comment:
Is this essentially recomputing the values that were used to create the haloArrays to begin with?  If so, could those values be combined in some way?  If not (maybe one is source, the other dest), could/should they be set up next to each other to show their relationship instead?

Reply:
This was simplified in the refactor to be based on the `localIndices` variable which I think is much more intuative.